src/domains/ValuesDomain.js

```diff
@@ -42,6 +69,185 @@ export default class ValuesDomain {
+      } else if (op === "/") {
+        // The sign of the result is positive if both operands have the same sign, negative if the operands have different signs.
+        // Division of an infinity by an infinity results in NaN.
+        // Division of an infinity by a zero results in an infinity. The sign is determined by the rule already stated above.
+        // Division of an infinity by a nonzero finite value results in a signed infinity. The sign is determined by the rule already stated above.
+        // Division of a finite value by an infinity results in zero. The sign is determined by the rule already stated above.
+        // Division of a zero by a zero results in NaN; division of zero by any other finite value results in zero, with the sign determined by the rule already stated above.
+        // Division of a nonzero finite value by a zero results in a signed infinity. The sign is determined by the rule already stated above.
+        return new NumberValue(realm, lnum / rnum);
+      } else if (op === "*") {
+        // The sign of the result is positive if both operands have the same sign, negative if the operands have different signs.
+        // Multiplication of an infinity by a zero results in NaN.
+        // Multiplication of an infinity by an infinity results in an infinity. The sign is determined by the rule already stated above.
+        // Multiplication of an infinity by a finite nonzero value results in a signed infinity. The sign is determined by the rule already stated above.
+        return new NumberValue(realm, lnum * rnum);
+      }
+    } else if (op === "!==") {
+      return new BooleanValue(realm, !StrictEqualityComparison(realm, lval, rval));
+    } else if (op === "===") {
+      return new BooleanValue(realm, StrictEqualityComparison(realm, lval, rval));
+    } else if (op === "!=") {
+      return new BooleanValue(realm, !AbstractEqualityComparison(realm, lval, rval));
+    } else if (op === "==") {
+      return new BooleanValue(realm, AbstractEqualityComparison(realm, lval, rval));
+    } else if (op === "&" || op === "|" || op === "^") {
+      // ECMA262 12.12.3
+
+      // 5. Let lnum be ? ToInt32(lval).
+      let lnum: number = ToInt32(realm, lval);
+
+      // 6. Let rnum be ? ToInt32(rval).
+      let rnum: number = ToInt32(realm, rval);
+
+      // 7. Return the result of applying the bitwise operator @ to lnum and rnum. The result is a signed 32 bit integer.
+      if (op === "&") {
+        return new NumberValue(realm, lnum & rnum);
+      } else if (op === "|") {
+        return new NumberValue(realm, lnum | rnum);
+      } else if (op === "^") {
+        return new NumberValue(realm, lnum ^ rnum);
+      }
+    } else if (op === "in") {
+      // ECMA262 12.10.3
+
+      // 5. If Type(rval) is not Object, throw a TypeError exception.
+      if (!(rval instanceof ObjectValue)) {
+        throw realm.createErrorThrowCompletion(realm.intrinsics.TypeError);
```

Comment:
Doesn't this object allocation and field initialization create some entries that are tracked by the realm? Just pointing it out, because the `catch`-logic in `binaryOp` seems to assume that `computeBinary` is pure. I'd prefer to make `computeBinary` truly pure.

Reply:
Good catch. This should throw a FatalError.